src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/CrossChainTransferStore.cs

```diff
@@ -177,155 +204,234 @@ private HashHeightPair TipToChase()
         /// within the wallet.
         /// </summary>
-        public async Task SanityCheckAsync()
+        /// <param name="crossChainTransfers">The transfers to check. If not supplied then all partial and fully signed transfers are checked.</param>
+        private void SanityCheck(ICrossChainTransfer[] crossChainTransfers = null)
         {
-            this.logger.LogTrace("()");
+            Recipient.FederationWallet wallet = this.federationWalletManager.GetWallet();
 
-            // Ensure that the store is synchoronized with the wallet.
-            if (await this.SynchronizeAsync())
+            if (crossChainTransfers == null)
             {
-                Recipient.FederationWallet wallet = this.federationWalletManager.GetWallet();
+                crossChainTransfers = Get(
+                    this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
+                        this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            }
 
-                using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            var tracker = new StatusChangeTracker();
+            foreach (CrossChainTransfer partialTransfer in crossChainTransfers)
+            {
+                // Verify that the transaction input UTXO's have been reserved by the wallet.
+                if (!SanityCheck(partialTransfer.PartialTransaction, wallet, partialTransfer.Status == CrossChainTransferStatus.FullySigned))
                 {
-                    dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
+                    tracker.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
+                }
+            }
 
-                    CrossChainTransfer[] partialTransfers = this.Get(dbreezeTransaction,
-                        this.depositsIdsByStatus[CrossChainTransferStatus.Partial].Union(
-                            this.depositsIdsByStatus[CrossChainTransferStatus.FullySigned]).ToArray());
+            if (tracker.Count == 0)
+                return;
+
+            using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
+            {
+                dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
 
-                    foreach (CrossChainTransfer partialTransfer in partialTransfers)
+                try
+                {
+                    foreach (KeyValuePair<ICrossChainTransfer, CrossChainTransferStatus?> kv in tracker)
                     {
-                        if (!SanityCheck(partialTransfer.PartialTransaction, wallet))
-                        {
-                            this.SetTransferStatus(partialTransfer, CrossChainTransferStatus.Rejected);
-                            await this.PutTransferAsync(dbreezeTransaction, partialTransfer);
-                        }
+                        this.PutTransfer(dbreezeTransaction, kv.Key);
```

Comment:
This method is persisting the cross-chain transfer - unless I read this code I wouldn't have know this method `SanityCheck` was doing other types of operations.  Maybe separate the logic here ? (if there's time).

Reply:
`SanityCheck` sets invalid transfers to `Rejected`. As part of the same logic such changes are persisted to the DB.

Reply:
Cool, seems like this method could be called `ValidateTransfers` or `ValidateCrossChainTransfers` .

Reply:
Renamed to `ValidateTransaction` and `ValidateCrossChainTransfers`.